protocols/kad/src/behaviour.rs

```diff
@@ -836,229 +922,232 @@ where
+            Quorum::Three => 3,
+            Quorum::Majority => total / 2 + 1,
+            Quorum::All => total,
+            Quorum::N(n) => usize::min(total, n.get() as usize)
+        }
+    }
 }
 
-/// Output event of the `Kademlia` behaviour.
+//////////////////////////////////////////////////////////////////////////////
+// Events
+
+/// The events produced by the `Kademlia` behaviour.
+///
+/// See [`Kademlia::poll`].
 #[derive(Debug, Clone)]
-pub enum KademliaOut {
-    /// We have discovered a node.
-    ///
-    /// > **Note**: The Kademlia behaviour doesn't store the addresses of this node, and therefore
-    /// >           attempting to connect to this node may or may not work.
+pub enum KademliaEvent {
+    /// The result of [`Kademlia::bootstrap`].
+    BootstrapResult(BootstrapResult),
+
+    /// The result of [`Kademlia::get_closest_peers`].
+    GetClosestPeersResult(GetClosestPeersResult),
+
+    /// The result of [`Kademlia::get_providers`].
+    GetProvidersResult(GetProvidersResult),
+
+    /// The result of periodic queries initiated by [`Kademlia::start_providing`].
+    AddProviderResult(AddProviderResult),
+
+    /// The result of [`Kademlia::get_record`].
+    GetRecordResult(GetRecordResult),
+
+    /// The result of [`Kademlia::put_record`].
+    PutRecordResult(PutRecordResult),
+
+    /// A new peer in the DHT has been discovered during an iterative query.
     Discovered {
-        /// Id of the node that was discovered.
+        /// The identifier of the discovered peer.
         peer_id: PeerId,
-        /// Addresses of the node.
+        /// The known addresses of the discovered peer.
         addresses: Vec<Multiaddr>,
-        /// How the reporter is connected to the reported.
+        /// The connection status of the reported peer, as seen by the local
+        /// peer.
         ty: KadConnectionType,
     },
 
-    /// A node has been added to a k-bucket.
-    KBucketAdded {
-        /// Id of the node that was added.
-        peer_id: PeerId,
-        /// If `Some`, this addition replaced the value that is inside the option.
-        replaced: Option<PeerId>,
+    /// A peer in the DHT has been added to the routing table.
+    RoutingUpdated {
+        /// The ID of the peer that was added to a bucket in the routing table.
+        new_peer: PeerId,
+        /// The ID of the peer that was evicted from the routing table to make
+        /// room for the new peer, if any.
+        old_peer: Option<PeerId>,
     },
+}
 
-    /// Result of a `FIND_NODE` iterative query.
-    FindNodeResult {
-        /// The key that we looked for in the query.
-        key: PeerId,
-        /// List of peers ordered from closest to furthest away.
-        closer_peers: Vec<PeerId>,
-    },
+/// The result of [`Kademlia::get_record`].
+pub type GetRecordResult = Result<GetRecordOk, GetRecordError>;
+
+/// The successful result of [`Kademlia::get_record`].
+#[derive(Debug, Clone)]
+pub struct GetRecordOk {
```

Comment:
Again, this is too much typing in my opinion.
Having to write things such as `GetRecordResult(Ok(GetRecordOk { records }))` is quite annoying to figure out when you're not the one who wrote the Kademlia code.

Reply:
While you can, you would usually not mention the inner `Ok` struct type in the pattern match though, i.e.

`GetRecordResult(Ok(result)) => { ... result.records ... }`

which surely isn't much worse than

`GetRecordResult(Ok(records)) => { ... records ... }`

or 

`GetRecordResult { records } => { ... records ... }`

the last of which does not even permit failure and is thus ruled out from my perspective. I did consider a single, unified error enum emitted by a single error event, but the only advantage I could see of that would be less effort on the implementors side (me) for loss of precision (i.e. any operation can supposedly yield any error, which isn't true) and there would then also need to be a way to identify what type of query actually produced the error. So I gladly took on the minor additional keyboard typing effort. 

If you simply suggest to drop `XyzOk` structs that (right now) just contain a single field, that seems possible but I like the consistency and extensibility of this API, where all DHT operations produce a `Result` with a struct for `Ok` and an enum for `Error`, making it easy to add new fields or variants specific to each type of query.